dodo.py

```diff
@@ -0,0 +1,325 @@
+##
+# This file contains various build steps used to make sure the data the Hedy application
+# uses is up-to-date.
+#
+# Functions that start with the name `task_` create a "task", which runs one or more
+# "actions" (shell commands or Python functions). Tasks can have file dependencies and
+# file outputs, as well as other task dependencies. If none of the files that a task
+# depends on have changed since the last time it ran, the task will be skipped. That's
+# good for speed!
+#
+# Keep the following in mind:
+# - You can specify commands as a string `'mkdir dir'` or as an array `['mkdir', 'dir']`.
+#   The second case doesn't allow shell features, but is safer in case of variables.
+# - `file_dep` requires a list of source files, but you can use `glob('*.yaml')`
+#   to have Python list them. That way you don't have to keep the list of files up-to-date
+#   by hand.
+# - It's important to be complete in the list of inputs that go into a build step. You
+#   have to list all input files, or the step will not accurately re-run when you change
+#   one of the forgotten dependencies.
+# - For eexample: include the script itself in the list of `file_dep`s as well;
+#   otherwise, even if you change the script the task won't re-run.
+
+import os
+from os import path
+from glob import glob
+import sys
+
+
+# The current Python interpreter, use to run other Python scripts as well
+python3 = sys.executable
+
+# When run without arguments, run deploy
+DOIT_CONFIG = {'default_tasks': ['deploy']}
+
+
+def task_tailwind():
+    """Build Tailwind.
+
+    We will automatically switch between DEV and PROD mode depending
+    on where we run.
+    """
+    prod = os.getenv('DYNO') is not None
+
+    if prod:
+        script = 'build-tools/heroku/tailwind/generate-prod-css'
+        target = 'static/css/generated.css'
+    else:
+        script = 'build-tools/heroku/tailwind/generate-development-css'
+        target = 'static/css/generated.full.css'
+
+    return dict(
+        file_dep=[
+            *glob('templates/**/*.html'),
+            *glob('main/**/*.md'),
+            *glob('content/**/*.md'),
+            *glob('static/js/*.ts'),
+            script,
+        ],
+        title=lambda _: 'Generate Tailwind CSS',
+        actions=[
+            [script],
```

Comment:
so these files would still run with bash? I personally don't mind bash as a build-tool at all.

Reply:
I'm wondering since this PR aims at using `doit` as the primary build-tool. But I'm guessing you kept this file just because we may need to run it locally; e.g., watching changes!

Reply:
> * `doit` will take care of dependencies: run tasks that depend on each other in the right order, won't run them twice, and won't run them if it's not necessary

Wanted to investigate whether it does that. So, i created this dummy function:
```python
# Create a simple task
def task_copy_file():
    return dict(
      title=lambda _: "copy file into another",
      actions=['cp "dev_database.json" "test here.txt"'],
      file_dep=['dev_database.json'],
      targets=['test here.txt']
      )
```

Then ran: `doit copy_file`. Notably, it did two things:
1. it ran the task that you created: `devdb` and `copy_file`. Probably because the file dep in copy_file is in the target of devdb!
2. i ran it again after a few minutes, and as you explained, it actually did not rerun the tasks. Really nice!!

Later, i removed `.doit.db`, and reran it, boths tasks were fired. So, i do believe we need to include it to git, otherwise all tasks will be executed!?

Reply:
> Probably because the file dep in copy_file is in the target of devdb!

Yes! Exactly, it traces task dependencies through declared file dependencies. In this case, that's probably not what we want to happen, so that might be something we need to be careful with...

> `.doit.db` [...] we need to include it to git, otherwise all tasks will be executed!?

I've been thinking about this a little, and I think it's safer to leave the database out of git for now.

* First, the database is a binary file, and git doesn't deal well with binary files usually
* Second, especially because it's a binary file, and many PRs will make changes that cause changes to this file, we will nearly always have conflicts on this file when trying to merge any PR. That will get annoying very quickly.
* If we check this file in, it will also go to the server, and the server will also skip build steps based on this file. That MAY be fine, but it may also become the cause of hard-to-diagnose bugs in the future, if any bugs slip in (if we make mistakes in dependency declarations on the tasks, for example).

The only reason to check in the database file would be to save running tasks across machines. I.e., I run the task, and then you don't have to anymore. That's nice, but it's not the end of the world if you have run the tasks *once*, and can skip them after.

All that is to say, I think the benefit is not massive, and the annoyances and risks are real, so I don't think checking in the database file would be worth it.